src/internal/pfsdb/v2.4.0.go

```diff
@@ -0,0 +1,171 @@
+package pfsdb
+
+import (
+	"context"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/pachyderm/pachyderm/v2/src/pfs"
+
+	col "github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+)
+
+// returns collections released in v2.4.0 - specifically the projects collection
+// DO NOT MODIFY THIS FUNCTION
+// IT HAS BEEN USED IN A RELEASED MIGRATION
+func CollectionsV2_4_0() []col.PostgresCollection {
+	return []col.PostgresCollection{
+		col.NewPostgresCollection(projectsCollectionName, nil, nil, nil, nil),
+	}
+}
+
+func MigrateProjectV2_4_0(p *pfs.Project) *pfs.Project {
+	if p == nil || p.Name == "" {
+		return &pfs.Project{Name: "default"}
+	}
+	return p
+}
+
+func MigrateRepoV2_4_0(r *pfs.Repo) *pfs.Repo {
+	r.Project = MigrateProjectV2_4_0(r.Project)
+	return r
+}
+
+func migrateRepoInvoV2_4_0(r *pfs.RepoInfo) *pfs.RepoInfo {
+	r.Repo = MigrateRepoV2_4_0(r.Repo)
+	for i, b := range r.Branches {
+		r.Branches[i] = MigrateBranchV2_4_0(b)
+	}
+	return r
+}
+
+func MigrateBranchV2_4_0(b *pfs.Branch) *pfs.Branch {
+	b.Repo = MigrateRepoV2_4_0(b.Repo)
+	return b
+}
+
+func migrateBranchInfoV2_4_0(b *pfs.BranchInfo) *pfs.BranchInfo {
+	b.Branch = MigrateBranchV2_4_0(b.Branch)
+	if b.Head != nil {
+		b.Head = MigrateCommitV2_4_0(b.Head)
+	}
+	for i, bb := range b.Provenance {
+		b.Provenance[i] = MigrateBranchV2_4_0(bb)
+	}
+	for i, bb := range b.DirectProvenance {
+		b.DirectProvenance[i] = MigrateBranchV2_4_0(bb)
+	}
+	for i, bb := range b.Subvenance {
+		b.Subvenance[i] = MigrateBranchV2_4_0(bb)
+	}
+	return b
+}
+
+func MigrateCommitV2_4_0(c *pfs.Commit) *pfs.Commit {
+	c.Branch = MigrateBranchV2_4_0(c.Branch)
+	return c
+}
+
+func migrateCommitInfoV2_4_0(c *pfs.CommitInfo) *pfs.CommitInfo {
+	c.Commit = MigrateCommitV2_4_0(c.Commit)
+	if c.ParentCommit != nil {
+		c.ParentCommit = MigrateCommitV2_4_0(c.ParentCommit)
+	}
+	for i, cc := range c.ChildCommits {
+		c.ChildCommits[i] = MigrateCommitV2_4_0(cc)
+	}
+	for i, bb := range c.DirectProvenance {
+		c.DirectProvenance[i] = MigrateBranchV2_4_0(bb)
+	}
+	return c
+}
+
+// MigrateV2_4_0 migrates PFS to be fully project-aware with a default project.
+// It uses some internal knowledge about how cols.PostgresCollection works to do
+// so.
+func MigrateV2_4_0(ctx context.Context, tx *pachsql.Tx) error {
+	if _, err := tx.ExecContext(ctx, `UPDATE pfs.commit_diffs SET commit_id = regexp_replace(commit_id, '^([-a-zA-Z0-9_]+)', 'default/\1') WHERE commit_id ~ '^[-a-zA-Z0-9_]+\.';`); err != nil {
```

Comment:
`commit_id ~ '^[-a-zA-Z0-9_]+\.'` why is there a `.` at the end?

Reply:
Not at the end — after the beginning. A pre-projects commit diff/total has a `commit_id` value which looks like `images.user@master=da4016a16f8944cba94038ab5bcc9933`; a post-projects commit diff/total has a `commit_id` which looks like `myproject/images.user@master=da4016a16f8944cba94038ab5bcc9933`.

The regexp above is meant to match only rows with a pre-projects–style `commit_id` and not touch post-projects–style rows. That is because prior to the default project name changing, commits in the default project were still identified without the project (e.g. as `images.user@master=da4016a16f8944cba94038ab5bcc9933` rather than `/images.user@master=da4016a16f8944cba94038ab5bcc9933`).

Reply:
Ah makes sense, thanks for the awesome explanation!

Reply:
No problem! It occurred to me that if it needs this much explanation, it should probably be a comment in the code … so now it is.